Makes attack() only handle unique overrides, move common code to use_weapon()

emmanuelbassil opened the pull request:
No changelog needed. Iirc pulled all the user facing stuff into separate PRs.

So this PR has alot of changes. Briefly; attack() now returns TRUE/FALSE and is part of the use_* chain. Attack() is only called if an item has the ITEM_TRY_TO_ATTACK flag; so if you are planning to create an attack() override you need to consciously enable the flag. Point of an attack() override is if you want special behavior to run before anything else when a mob clicks on another mob with an item, that you don't want all the use_*'s to handle. If attack() returns TRUE; further procs are not called.
**As a result, all attack() overrides should now be returning booleans, as per this PR.** 
If desired, we can just axe the flag completely and always have attack called first. It's a stub returning FALSE for any item without an override.

Another issue prior to this PR is that attack()'s parent called regular melee attack code. Use_weapon also called regular melee attack code. What's more annoying is that each method has features the other does not have. Up until now use_weapon did NOT process special weapon behavior (think a stun baton shocking), armor, miss chance, or even custom verbs. These were all handled by the series of procs called by attack; in order:
- Apply_hit_effect which plays a sound (already handled by use_weapon) and magnifies damage if HULK. 
- Hit_with_weapon which processes custom attack verbs. It also processes simple mobs reacting and random puddle of blood. 
- Standard_weapon_hit_effects which does alot of things, from joint dislocation with an item; sharp/edge/weapon flag processing, to knocking someone down.
- Then reaches apply_damage, a common point with use_weapon. Do note, use_weapon called **none** of the above behavior; resulting in lost functionality.

I did the following: 
1. Apply_hit_effect's sound code was moved to use_weapon. Apply_hit_effect only has the hulk stuff now, and is overrided by items that want special behavior after a successful hit. (Paddles, stun baton, etc.)
2. Moved hit_with_weapon's custom verb functionality to use_weapon's base proc. Moved ai_holder reacting to post_use_item. Axed the random puddle of blood for non-human functionality. 
3. I also moved all of standard_weapon_hit_effect's behavior to hit_with_weapon, since the proc didn't do anything useful anymore. This allows me to completely remove the proc standard_weapon_hit_effects. One less proc in this long chain.
4. Use_weapon **no longer** calls general_health_adjustment but calls apply_hit_effect instead and then goes down the above procs. This is because by calling general health adjustment, it was missing all this other behavior.
5. General health adjustment is no longer called by melee code; but goes straight to apply_damage. Useful functionality from general health adjustment was moved to apply_damage. This is because apply_damage takes armor penetration and damage flags as arguments whereas general health adjustment is more fit to do exactly what its name says: Directly adjust health.

Softly dependent on #33990, #34049 and #33992 since they also turn some attacks() into booleans that I did not do in this PR to ease the suffering, since there is an insane amount of modified files here.


List:
- [x] Go through all attack procs, make sure they now return booleans if processed or not. 
- [x] Go through procs called by attack, make sure their behavior is handled.
- [x] Some simple_animals use target_mob to attack, which is unused. Make them use ai_holder.target and test
- [x] Make holster and accessories use use_on instead of attackby
- [x] Remove non-unique behavior from attack() proc. Moved to use_weapon and simplify.
- [x] Delete airlock code under attackby that lets simple_animals break through doors. This is handled by ai_handler_combat.dm. 
- [ ] Go over show_in_messages var for simple_mobs, make sure everything is set correctly. (Will handle in separate PR)

SierraKomodo commented on code/_onclick/item_attack.dm:
```diff
@@ -406,36 +420,17 @@ avoid code duplication. This includes items that may sometimes act as a standard
-	if (hit_zone)
-		apply_hit_effect(subject, user, hit_zone)
-	return TRUE
+	else return FALSE
```
`return subject.use_weapon(src, user)`

SierraKomodo replied:
This isn't the section of code I clicked on. Thanks, github.

SierraKomodo commented on code/modules/clothing/under/accessories/_accessory.dm:
```diff
@@ -106,6 +106,12 @@
 	else
 		dropInto(loc)
 
+/obj/item/clothing/accessory/use_on(obj/target, mob/user)
+	if (istype(target, /obj/item/clothing))
+		var/obj/item/clothing/clothes = target
+		if (clothes.attempt_attach_accessory(src, user))
+			return TRUE
+	return ..()
```
This looks like it's identical to the `/obj/item/use_on()` parent.

SierraKomodo commented on code/modules/mob/living/simple_animal/defense.dm:
```diff
@@ -61,13 +61,11 @@
-	if (result && ai_holder)
-		ai_holder.react_to_attack(user)
+	//If an attack is successful; then react. Call parent before in this case, not after.
+	if (..())
+		ai_holder?.react_to_attack(user)
 		return TRUE
-
-	return ..()
+	else return FALSE
```
This can be moved to `post_use_item`:

```dm
/mob/living/simple_animal/post_use_item(obj/item/tool, mob/living/user, interaction_handled, use_call, click_params)
	..()
	if (interaction_handled && ai_holder && (use_call == "attack" || use_call == "weapon"))
		ai_holder.react_to_attack(user)
```

juxjux9930 commented on code/_onclick/other_mobs.dm:
```diff
@@ -196,14 +196,15 @@
-			custom_emote(1,"[friendly] [A]!")
+			custom_emote(VISIBLE_MESSAGE, "[friendly] [A]!")
 			return
 		if(ckey)
 			admin_attack_log(src, A, "Has attacked its victim.", "Has been attacked by its attacker.")
 	if(a_intent == I_HELP)
```
```suggestion
		if (ckey)
			admin_attack_log(src, A, "Has attacked its victim.", "Has been attacked by its attacker.")
	if (a_intent == I_HELP)
```
Might as well, while you're here

juxjux9930 commented on code/_onclick/other_mobs.dm:
```diff
@@ -196,14 +196,15 @@
 	setClickCooldown(DEFAULT_ATTACK_COOLDOWN)
 	if(istype(A,/mob/living))
 		if(!get_natural_weapon() || a_intent == I_HELP)
```
```suggestion
	if (istype(A,/mob/living))
		if (!get_natural_weapon() || a_intent == I_HELP)
```

juxjux9930 commented on code/game/gamemodes/cult/cult_items.dm:
```diff
@@ -7,10 +7,11 @@
+	attack_verb = list("attacked", "slashed", "stabbed", "sliced", "teared", "ripped", "diced", "cut")
 
 /obj/item/melee/cultblade/attack(mob/living/M, mob/living/user, target_zone)
 	if(iscultist(user))
```
```suggestion
	if (iscultist(user))
```
pad your contributions with this one, simple trick

juxjux9930 commented on code/game/gamemodes/cult/cult_items.dm:
```diff
@@ -38,7 +39,7 @@
+	return TRUE
 
 /obj/item/melee/cultblade/pickup(mob/living/user as mob)
 	if(!iscultist(user))
```
```suggestion
	if (!iscultist(user))
```

juxjux9930 commented on code/game/objects/items.dm:
```diff
@@ -283,6 +284,13 @@
 			else if(S.can_be_inserted(src, user))
```
```suggestion
			else if (S.can_be_inserted(src, user))
```

emmanuelbassil committed:
Expands use_weapon, allows simplemobs to use it

Fixes

Fixes

Auto stash before rebase of "refs/heads/grabs"
Auto stash before rebase of "refs/heads/grabs"
Auto stash before rebase of "refs/heads/grabs"
Auto stash before rebase of "refs/heads/grabs"

emmanuelbassil committed:
Makes attacks() binary, adds item_flag_try_attack


More fixes


Fixes


Auto stash before rebase of "refs/heads/grabs"

emmanuelbassil committed:
More Attack() booleanization

Spookerton commented:
mis-aimed fix hint from https://github.com/Baystation12/Baystation12/pull/34036

emmanuelbassil commented:
Ready for review on my end.
Pending merging #33992 and #34049

emmanuelbassil commented:
This is ready.
Removed SHOULD_CALL_PARENT(TRUE) from attack and its children that had it set to FALSE; no real need for it.

SierraKomodo commented on code/_onclick/item_attack.dm:
```diff
@@ -281,24 +280,41 @@ avoid code duplication. This includes items that may sometimes act as a standard
 			user.visible_message(
-				SPAN_WARNING("\The [user] hits \the [src] with \a [weapon], but it bounces off!"),
-				SPAN_WARNING("You hit \the [src] with \the [weapon], but it bounces off!"),
+				SPAN_WARNING("\The [user] hit \the [src] [weapon_mention], but it bounced off!"),
+				SPAN_WARNING("You hit \the [src] [weapon_mention], but it bounced off!"),
 				exclude_mobs = list(src)
```
Present tense is preferred for messages.

SierraKomodo commented:
Nothing immediately stands out to me, good for a test merge.

emmanuelbassil committed:
Make attack() a stub

Fixes

Auto stash before rebase of "refs/heads/grabs"
Auto stash before rebase of "refs/heads/grabs"
Fixes

Auto stash before rebase of "refs/heads/grabs"

Fixes


Auto stash before rebase of "refs/heads/grabs"
Fixes

SierraKomodo commented:
Test merge is live.

emmanuelbassil committed:
Fixes

emmanuelbassil committed:
Change ..() to FALSE


Fixes

emmanuelbassil commented:
Fixed bugs from test merge.
Flash, extinguisher, snacks, and medical stack items did not have an !ismob check preventing them from being stored/used.

Fixed bug that didn't allow ATK to be used in surgery. Attack() now returns FALSE if there's a possible surgical step for the ATK, which then proceeds to use_tool where the surgery is actually handled.

Removed all the ..() since there's no need for almost any of the attack overrides (except for pens and medical stacks iirc) to call a parent. Also made the default return value for almost all overrides FALSE; which is neater than having an intended return false at the end of each override.

Ready for test merge again.
